### PR TITLE
Convert dropdown task index to ES6

### LINKED ISSUE #4442: dropdown task with allowCreate - tab saves "Press enter for..." as answer
## Expected behavior
If dropdown task input allows custom answer (`allowCreate: true`) and I type "my custom answer" then press tab, I expect it to save that answer (or at minimum not do what the current behavior does)

## Current behavior
In scenario above, pressing tab saves "Press enter for my custom answer..." as answer.

## Steps to replicate
https://master.pfe-preview.zooniverse.org/projects/markb-panoptes/test-project-3-mb/classify?reload=0&workflow=3163

mentioning @trouille

### PR DESCRIPTION
**Pull request by mcbouslog**
Staging branch URL: https://dropdown-es6-index.pfe-preview.zooniverse.org/

Fixes #4442.

Describe your changes.

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/classifier/tasks/dropdown/index.jsx [hunk 1]
```diff
@@ -0,0 +1,292 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import Select from 'react-select';
+import GenericTask from '../generic';
+import DropdownEditor from './editor';
+
+export const DropdownSummary = (props) => {
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: `export function DropdownSummary(props {`

File: app/classifier/tasks/dropdown/index.jsx [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  const getOptionsKeys = () => {
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: `function getOptionsKeys() {`

File: app/classifier/tasks/dropdown/index.jsx [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+      } else if (answer.value === null) {
+        optionsKeys[id] = null;
+      } else {
+        optionsKeys[id] = `${optionsKeys[condition]};${answer.value}`;
+      }
+    });
+    return optionsKeys;
+  };
+
+  const getOptionLabel = (value, i) => {
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: This should be declared as a named function too.

I wonder if we can catch these with a linter rule?

File: app/classifier/tasks/dropdown/index.jsx [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+  getOptions(i) {
+    const select = this.props.task.selects[i];
+    const optionsKey = select.condition ? this.getOptionsKey(i) : '*';
+    const options = select.options[optionsKey];
+    return options || [];
+  }
+
+  getDisabledAttribute(i) {
+    const { selects } = this.props.task;
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: This function is a little hard to follow. Does it need a comment to clarify what the conditions are for a menu being disabled?

File: app/classifier/tasks/dropdown/index.jsx [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+  setDefaultValues() {
+    if (!this.props.annotation.value.length) {
+      const value = this.props.task.selects.map(() => ({ option: false, value: null }));
+      const newAnnotation = Object.assign({}, this.props.annotation, { value });
+      this.props.onChange(newAnnotation);
+    }
+  }
+
+  getOptionsKey(i, key) {
+    const { selects } = this.props.task;
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: This is another function that's hard to follow on first reading. I remember this one being hard to get right when working on this component before.

**Comment by mcbouslog** [comment]
added comments, unsure of how clarifying they are though, let me know what you think

File: app/classifier/tasks/dropdown/index.jsx [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    const selectedOptions = [];
+    this.props.annotation.value.map((annotation, i) => {
+      if (annotation.option) {
+        const [selected] = this.getOptions(i).filter(option => option.value === annotation.value);
+        selectedOptions[i] = selected;
+      } else {
+        selectedOptions[i] = { label: annotation.value, value: annotation.value };
+      }
+    });
+    return selectedOptions;
```
Comment (eatyourgreens) [inline app/classifier/tasks/dropdown/index.jsx]: It just occurred to me that this could be written as an actual array map. Something like:
```javascript
const selectedOptions = this.props.annotations.map((annotation, i) => {
  if (annotation.option) {
    const [selected] = this.getOptions(i).filter(option => option.value === annotation.value);
    return selected;
  } else {
    return { label: annotation.value, value: annotation.value };
  }
});
return selectedOptions;
```

**Comment by eatyourgreens** [comment]
Looks good to me. I think `optionsKeys` is always going to confuse me, but the comments made it clearer how they work. Just the one suggestion about the array map to generate the array of selected options. I don't think I properly understood maps when I first wrote that.

**Comment by mcbouslog** [comment]
Just retested on a few NfN workflows and appears to be working as usual/expected. I'll keep an eye on the NfN Talk boards once merged.
